sdk/python/feast/infra/utils/snowflake_udfs.py

```diff
@@ -0,0 +1,239 @@
+    )
+    return df
+
+
+"""
+CREATE OR REPLACE FUNCTION feast_snowflake_boolean_to_bool_proto(df BOOLEAN)
+  RETURNS BINARY
+  LANGUAGE PYTHON
+  RUNTIME_VERSION = '3.8'
+  PACKAGES = ('protobuf', 'pandas')
+  HANDLER = 'feast.infra.utils.snowflake_udfs.feast_snowflake_boolean_to_bool_boolean_proto'
+  IMPORTS = ('@feast_stage/feast.zip');
+"""
+# ValueType.BOOL = 7
+@vectorized(input=pandas.DataFrame)
+def feast_snowflake_boolean_to_bool_boolean_proto(df):
+    df = list(
+        map(
+            ValueProto.SerializeToString,
+            python_values_to_proto_values(df[0].to_numpy(), ValueType.BOOL),
+        )
+    )
+    return df
+
+
+"""
+CREATE OR REPLACE FUNCTION feast_snowflake_timestamp_ntz_to_unix_timestamp_proto(df TIMESTAMP_NTZ)
+  RETURNS BINARY
+  LANGUAGE PYTHON
+  RUNTIME_VERSION = '3.8'
+  PACKAGES = ('protobuf', 'pandas')
+  HANDLER = 'feast.infra.utils.snowflake_udfs.feast_snowflake_timestamp_ntz_to_unix_timestamp_proto'
+  IMPORTS = ('@feast_stage/feast.zip');
+"""
+# ValueType.UNIX_TIMESTAMP = 8
+@vectorized(input=pandas.DataFrame)
+def feast_snowflake_timestamp_ntz_to_unix_timestamp_proto(df):
+    df = list(
+        map(
+            ValueProto.SerializeToString,
+            python_values_to_proto_values(df[0].to_numpy(), ValueType.UNIX_TIMESTAMP),
+        )
+    )
+    return df
+
+
+"""
+CREATE OR REPLACE FUNCTION feast_serialize_entity_keys(names ARRAY, data ARRAY, types ARRAY)
+  RETURNS BINARY
+  LANGUAGE PYTHON
+  RUNTIME_VERSION = '3.8'
+  PACKAGES = ('protobuf', 'pandas')
+  HANDLER = 'feast.infra.utils.snowflake_udfs.feast_serialize_entity_keys'
+  IMPORTS = ('@feast_stage/feast.zip')
+"""
+# converts 1 to n many entity keys to a single binary for lookups
+@vectorized(input=pandas.DataFrame)
+def feast_serialize_entity_keys(df):
+    join_keys = create_entity_dict(df[0].values[0], df[2].values[0])
+
+    df = pandas.DataFrame.from_dict(
+        dict(zip(df[1].index, df[1].values)), orient="index", columns=df[0].values[0]
+    )
+
+    proto_values_by_column = {
+        column: python_values_to_proto_values(df[column].to_numpy(), value_type)
+        for column, value_type in list(join_keys.items())
+    }
+
+    serialized_entity_keys = [
+        serialize_entity_key(
+            EntityKeyProto(
+                join_keys=join_keys,
+                entity_values=[proto_values_by_column[k][idx] for k in join_keys],
+            )
+        )
+        for idx in range(df.shape[0])
+    ]
+    return serialized_entity_keys
+
+
+"""
+CREATE OR REPLACE FUNCTION feast_entity_key_proto_to_string(names ARRAY, data ARRAY, types ARRAY)
+  RETURNS BINARY
+  LANGUAGE PYTHON
+  RUNTIME_VERSION = '3.8'
+  PACKAGES = ('protobuf', 'pandas')
+  HANDLER = 'feast.infra.utils.snowflake_udfs.feast_entity_key_proto_to_string'
+  IMPORTS = ('@feast_stage/feast.zip')
+"""
+# converts 1 to n many entity keys to a single binary for lookups
+@vectorized(input=pandas.DataFrame)
+def feast_entity_key_proto_to_string(df):
```

Comment:
SNowflake passes an array of variants, which might have another datatype? How do this arrive in the python UDF? And does it matter? Does python_values_to_proto_values create the right proto as long as it knows ValueType?

https://docs.snowflake.com/en/sql-reference/data-types-semistructured.html#array

Reply:
cc @adchia